h2/src/main/org/h2/value/DataType.java

```diff
@@ -390,6 +397,23 @@
         }
     }
 
+    public static int registerUserType(Class<? extends ValueUserDefined> valueClass, String name) {
```

Comment:
All the following stuff looks suspicious. I'd prefer to avoid having any separate typeId for user defined values.